### ISSUE TITLE
False Positive 'TLS1.3 Server accepts all signature algorithms'

### ISSUE THREAD

**Issue by emanuelfc** [rbsec/sslscan#282]
## Summary

_sslscan_ may return a False Positive `'Server accepts all signature algorithms'` for TLS1.3.

## Example

Signature Algorithm =  `RSA+SHA256 (rsa_pkcs1_sha256)`

Not supported by `aws.amazon.com`.

```
openssl s_client -connect aws.amazon.com:443 -sigalgs "RSA+SHA256" -tls1_3
```

![image](https://user-images.githubusercontent.com/26048363/236208189-90625235-a408-4df8-bf33-f4a38ff10502.png)

However, _sslscan_ reports as `'Server accepts all signature algorithms'`.

```
sslscan --show-sigs --no-cipher-details --no-ciphersuites --no-compression --no-fallback --no-groupseartbleed --no-renegotiation --no-check-certificate --tls13 aws.amazon.com
```

![image](https://user-images.githubusercontent.com/26048363/236208341-c7917eac-82af-4bb5-93b2-1a5384c095c8.png)

## Explanation

Currently, _sslscan_ only verifies if the _ServerHello_ message was successfully received, and if so, then the given signature algorithm is considered to have been accepted by the server.

https://github.com/rbsec/sslscan/blob/b31459eb25324ddae52293b5ac46f2c86e8f7e51/sslscan.c#L6011-L6044

However, TLS1.3 now makes use of the _CertificateVerify_ message:

**TL;DR**: If the server does not support any of the signature algorithms offered by the client, it continues the handshake and may use a different signature algorithm.

> ...
> Servers MUST send this message when authenticating via a certificate.
> ...
> If the CertificateVerify message is sent by a server, the signature algorithm MUST be one offered in the client's "signature_algorithms" extension unless no valid certificate chain can be produced without unsupported algorithms (see [Section 4.2.3](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.3)).
> 
> https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.3

> "If the server cannot produce a certificate chain that is signed only via the indicated supported algorithms, then it SHOULD continue the handshake by sending the client a certificate chain of its choice that may include algorithms that are not known to be supported by the client. This fallback chain SHOULD NOT use the deprecated SHA-1 hash algorithm in general, but MAY do so if the client's advertisement permits it, and MUST NOT do so otherwise."
> 
> https://datatracker.ietf.org/doc/html/rfc8446#page-67

As such, it fails to obtain the _CertificateVerify_ message sent by the server and verify if the signature algorithm specified by the client was accepted, or if the server using a different one.

```
openssl s_client -connect aws.amazon.com:443 -sigalgs "RSA+SHA256" -msg -trace
```

![image](https://user-images.githubusercontent.com/26048363/236207879-2acb6416-d245-4a3d-9ab1-98957649d9c7.png)

## Remediation

When receiving a _CertificateVerify_ message from the server, verify if the signature algorithm matches the one sent by the client.

**Comment by dougch** [comment]
[This is an invalid assumpiton](https://github.com/rbsec/sslscan/blob/e0563b9acd4e32da4732cbdeb258e453638b1388/sslscan.c#L6025):
```
      /* If the server accepted our bogus signature ID, then we can conclude that it will accept all of them (and not test any further).  Some servers in the wild do this for some reason... */
```
Some clients lie about what signature algorithms they will accept.  If you remove  the break on line [6029](https://github.com/rbsec/sslscan/blob/e0563b9acd4e32da4732cbdeb258e453638b1388/sslscan.c#L6029), 

These pass against aws.amazon.com
```
TLSv1.3  ed25519
TLSv1.3  ed448
TLSv1.2  ed25519
TLSv1.2  ed448
```
